File: 0001-0500/0032/0032.longest-valid-parentheses.cs
/*
 * @lc app=leetcode id=32 lang=csharp
 *
 * [32] Longest Valid Parentheses
 */

// @lc code=start
public class Solution {
    public int LongestValidParentheses(string s) {
        int max = 0;
        int[] dp = new int[s.Length];
        for(int i = 1; i < s.Length; i++) {
            if(s[i] == ')') {
                if(s[i - 1] == '(') {
                    dp[i] = (i >= 2 ? dp[i - 2] : 0) + 2;
                } else if(i - dp[i - 1] > 0 && s[i - dp[i - 1] - 1] == '(') {
                    dp[i] = dp[i - 1] + ((i - dp[i - 1]) >= 2 ? dp[i - dp[i - 1] - 2] : 0) + 2;
                }
                max = Math.Max(max, dp[i]);
            }
        }
        return max;
    }
}
// @lc code=end

File: 0001-0500/0032/0032.longest-valid-parentheses.js
/*
 * @lc app=leetcode id=32 lang=javascript
 *
 * [32] Longest Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function (s) {
    var stack = [];
    var max = 0;
    var start = 0;
    for (var i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i);
        } else {
            if (stack.length === 0) {
                start = i + 1;
            } else {
                stack.pop();
                if (stack.length === 0) {
                    max = Math.max(max, i - start + 1);
                } else {
                    max = Math.max(max, i - stack[stack.length - 1]);
                }
            }
        }
    }
    return max;
};
// @lc code=end

